cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js

```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+var server = require('server');
+
+/* Script Modules */
+var shippingHelper = require('*/cartridge/scripts/checkout/shippingHelpers');
+var COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
+var collections = require('*/cartridge/scripts/util/collections');
+var cartHelpers = require('*/cartridge/scripts/cart/cartHelpers');
+
+/* Models */
+var CartModel = require('*/cartridge/models/cart');
+
+/* API Includes */
+var BasketMgr = require('dw/order/BasketMgr');
+var ProductMgr = require('dw/catalog/ProductMgr');
+var PromotionMgr = require('dw/campaign/PromotionMgr');
+var StringUtils = require('dw/util/StringUtils');
+var Transaction = require('dw/system/Transaction');
+var URLUtils = require('dw/web/URLUtils');
+
+
+/**
+ * KlaviyoRecreate-Cart : The KlaviyoRecreate-Cart route rebuilds a cart page based on a query containing an ENCODED array of objects with product IDs, product quanty, the product's selected options within URL / endpoint.
+ * A DECODED query would be structured as:  <baseURL>/Cart-Recreate?items=[ {"productID": ProdID, "quantity": QTY, "options": [ {lineItemText, optionID, optionValueID} ]} ]
+ * @function
+ * @memberof Cart
+ * @param {querystringparameter} - items - JSON containing product Ids, qty, childProducts, options
+ * @param {renders} - isml
+ * @param {serverfunction} - get
+ */
+server.get('Cart', function (req, res, next) {
+    var currentBasket = BasketMgr.getCurrentOrNewBasket();
+    try {
+        var items = req.querystring.items ? JSON.parse(StringUtils.decodeBase64(req.querystring.items)) : null;
+    } catch (error) {
+        res.setStatusCode(500);
+        res.json({
+            error: true,
+            errorMessage: `ERROR - ${error.message}. Please check the encoded obj for any unexpected chars or syntax issues.`,
+        });
+        return next();
+    }
+
+    if (!currentBasket) {
+        res.setStatusCode(500);
+        res.json({
+            error: true,
+            errorMessage: `ERROR - Current Basket is: ${currentBasket}. Please check the Current Basket and try again.`,
+            redirectUrl: URLUtils.url('Cart-Show').toString()
+        });
+        return next();
+    };
+
+    // Clean the basket to prevent product duplication on page refresh
+    if (currentBasket && currentBasket.productQuantityTotal > 0) {
+        for (let i = 0; i < currentBasket.productLineItems.length; i++) {
+            let lengthAtIteration = currentBasket.productLineItems.length;
```

Comment:
The logic has me a bit confused. I think you could accomplish the goal here by simply stepping backwards through the cart instead using (let i = currentBasket.productLineItems.length-1; i >= 0; i--) or similar? Incrementing and decrementing i makes this a bit hard to follow and debug.

Reply:
I'm also wondering why this logic was not needed in recreateCartHelpers.js addProductToCart. If cart.object addProductToCart is the raw SFCC LineItemCtnr API object, could you create an emptyCart(lineItemCtnr) helper function that could be used by both SFRA and sitegen/controllers? Seems like the LineItemCtnr.allProductLineItems used in addProductToCart could also be used here in lieu of LineItemCtnr.productLineItems to avoid the PromotionMgr call at line 69. Not sure why the shipment removal logic is needed here, but not in addProductToCart, unless that's already included in the sitegen removeProductLineItem logic (not viewable in this PR).

Reply:
Thanks - have reworked following our chat on 04.24. Changes related to this are now available in the following PR:

- https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/19